app/utility/base_object.py

```diff
@@ -29,7 +29,7 @@ def match(self, criteria):
             return self
 
     def update(self, field, value):
-        if (value or type(value) == list) and (value != self.__getattribute__(field)):
+        if (value is not None) and (value != self.__getattribute__(field)):
```

Comment:
This change is a little worrying to me in case there's some part of caldera or a plugin somewhere that relies on the conditional here.  I much prefer checking it for None though.  

What's the problem that this fixes?  And, how thoroughly have you put caldera through it's paces after making the change?

Reply:
To echo @unkempthenry's thoughts - functionally, these may seem identical, but there's a possibility of an edge case with an empty list that worries me a little bit. I'm not saying that the edge case necessarily exists (I don't quite know all the uses for this function), but the possibility is there unless we've thoroughly tested for it.

Reply:
The `has_repeatable_abilities` value is a boolean, so if you had a profile with repeatable abilities (`has_repeatable_abilities = true`) and then removed the repeatable abilities and tried to save the profile, the `has_repeatable_abilities` value will not update because of the logic `value or type(value) == list`, since it will be `false` on the first case and also not a list. 

I'm not certain if there was a specific case or reason for how we previously had it, but checking for `not None` appears to cover the first check and the `type(value) == list` as well. To my knowledge, there aren't any other cases that update a boolean attribute to False.

Reply:
Hmm... Yeah good point about booleans.  I wonder what problem this conditional was actually trying to solve -- seems like kind of a hack -- why not just always update with the value ?

Reply:
git historian work: 

here's the commit where it changed to accept lists: 

https://github.com/mitre/caldera/pull/1455/commits/7a5bc34e6926ea8cc97a789cf1d51a1a822765ca#diff-1f946083235cdc5323c717bebe81445f0dc08a311b7b686e5875595c62e1b61bL27

Reply:
Even if we wanted to prevent some values -- I think this function should raise an exception in those cases instead of just silently failing...

Reply:
From a few quick tests, it appears that this logic prevents properties that aren't passed back from being overwritten to `None` while allowing empty lists to be updated properly (especially in the case of removing payloads from an ability). 

This is definitely used in the case for agents where [`heartbeat_modification`](https://github.com/mitre/caldera/blob/d878e65db22df9349256190d1bb087cd803f5936/app/objects/c_agent.py#L134) attempts to update anything that comes back in `kwargs`, but for the most part it's just the agent `paw` that is returned, so this logic prevents the other agent properties from getting overwritten to `None`. 

This also appears to be the case for saving abilities (`privilege`, `code`, `language`, `build_target` do not get overwritten to `None`) and adversaries (`tags`) as well, though I don't believe any of those properties are actually ever changed via `update`. There may be other cases too, but these are the cases I was able to identify immediately.

The way the logic is now, we aren't able to set properties to `None`. It's absolutely possible to add an `else` and raise an exception or log a DEBUG message when this happens, but the log will be flooded with the "attempted to set a property to none" every agent heartbeat. If that's acceptable, I'm fine with proceeding to do that. Thoughts @unkempthenry ?

Reply:
Alternatively, maybe figuring out a way to only update the values intentionally passed back instead of calling update on all properties that could possibly be passed back might be the route we want to take. That would resolve the issue of not being able to use `update` to set a property to `None`.

Reply:
Thanks for digging into it, and doing some testing @mchan143.  Can you add a docstring explaining that None values will be ignored? 

Raising an exception here would probably require refactoring every call to this function and that is definitely out of scope of this PR.  

Down the road, if we ever want to allow updating with None types without a heavy refactor we could probably do something like: 

```python
    def update(self, field, value, allow_nonetype=False):
        if not allow_nonetype and value is not None and value != self.__getattribute__(field):
            ...
```

But, I don't think we need to add it now.

Reply:
docstring added in https://github.com/mitre/caldera/pull/2019/commits/fdd27ccfd7def248f8b6b5cd98571b90d15edc20